### ISSUE TITLE
[NG] Implement a Tree-View Component

### ISSUE THREAD

**Issue by ariquinones** [ilanddev/ferui#43]
## We need to implement a side-navigation Tree-View component

_This tree-view component will be a hierarchical component that will show a visual representation of a parent-child relationship between its tree nodes._

Requirements:

- Element needs both horizontal and vertical scrolling (we will not know how many sub-directories exist nor how many levels deep it could go)
- Width and height of this element will be set forth by the user in configuration otherwise a default will be used
- Optional* - toggle whether to show non-folder (item) types
- Server Side - Discuss a lazy loading option when getting limited sub-directories (load a first batch and trigger more by scrolling or a load more option), can also be part of configuration
- Add loading animation
- Implement its interface to have:
  - Configuration: 
    - loading: boolean to show loading animation if data not yet available
    - serverSide: ServerSideDataInterface - optional, will let us know if component is to be used server-side and this interface will be the dataset for all server-side operations

      ```typescript
        // Bind to a method to request tree view data
        interface ServerSideDataInterface {
          getDirectories: (queryParams: ServerParams) => TreeViewInterface; // This will return new tree-view data values
        }
        interface ServerParams {
          limit: number,
          offset: number,
          marker?: string,
          sort?: string
        }
      ```
    - width: optional, if a user wants to set the width, otherwise it will auto fit its container
    - height: optional, if a user wants to set the height, otherwise it will auto fit its container
    - entitiesToHide: TreeViewDirectoryTypeEnum - optional should the user want to show only certain types of entities in tree-view in accordance to the tree view interface object
    - directories: TreeViewInterface{} - data object as input to display in tree-view 
      ```json
        { 
          "root": {
            "name": "Root folder",
            "sub_directories": [{
              "name": "first child directory",
              "type": "Directory",
              "expanded": "false",
              "selected": "false",
              "sub_directories": []
              }]
           }
        }
      ```

  - Methods:
    - addNode - add node to tree-view
    - contain - search for a certain entity(search purposes)
    - viewMore - when only showing certain amount of nodes
    - viewLess - show certain amount of nodes at a time
  - Events:
     - selected - subscribe to an entity being selected
     - collapse - subscribe to a sub-dir being collapsed and its children hide
     - expand - subscribe to a sub-dir being expanded and its children show
     - change - subscribe to configuration data update

## Example
![Screen Shot 2019-09-03 at 2 32 15 PM](https://user-images.githubusercontent.com/16388180/64220862-15e95c80-ce88-11e9-858f-234bf28d121b.png)

**Comment by JoMas971** [comment]
We won't actually do this exact component in FerUI, only the tree view component (the left section) and the breadcrumb components. The datagrid is already part of FerUI. 
This all-in-one component will be done within the console side but it will use all the datagrid, breadcrumb, and the tree-view components from FerUI.
So we'll need to think of all implications a tree-view component can have like Cory said :
- load it's hierarchy from server side or in memory. In all cases, we need to create an interface for the dataset. 
- Trigger events that any other components can listen to (not only the datagrid). 
- Its size can be set to auto (the size depends on the number of branches/leafs) or set to a specific value in any possible units (px, em, rem, %...). A scrollbar can appear in some cases of course. 
- We also need to know if an element is a branch, a leaf or a root in order to be able to design it easily.
- The routing will not be tied to this component. Instead, it would be up to the dev to bound events from the treeview to their own rooting system (FerUI need to be compatible with any other frameworks, not only Angular so we can't depend on a component like the angular router). 

It should be a generic component, not tied to the datagrid only. I'll use it as the left menu for the demo page ;-). 

And, we will create a breadcrumb component as well in FerUI later on :-).

**Comment by ariquinones** [comment]
@JoMas971 yup. Was misunderstanding the component at first. I edited my comment to take out all Datagrid references. Making a generic component to be used anywhere and communicate with any other component now.
My first comment is just a start, I'll be adding more details as I can think of all we will need from this component (screenshot of design is also updated, only tree nodes are shown)

**Comment by JoMas971** [comment]
@ariquinones much better now indeed :-D
I just have some comments about the interface you want to implement : 
- The `size` attribute is not enough. We would need both `width` and `height` instead and those options would be optional. By default, if we don't specify any width or height, the component should take all the space available for it. It will be up to the dev to wrap that component with whatever container he want.
- I think the `Directories` option should have its own interface and not be Object only. Because we need to be able to localise the roots, branches and leaf from the object representation. We need to decide how we represent the parent and the child. Should we use a `children` attribute or use another one etc... That's why i think we should create a specific interface for the `Directories` option.
- `EntitiesToShow` will depend on the interface you'll create for the `Directories` option. and i think we could also have an array (for non-hierarchical representation or if we only have one layer but still want to use the treeView component and all it's cool features). So it couldn't be an array of string. Instead we should use an interface. 
- I personally don't like the `LoadMore` function. I rather like to have a dataset for all server-side operations. Like i did for the datagrid. Because you could have filtering, sorting, pagging and many other possible options for server-side calls. And having just a `LoadMore` option doesn't look enough for me. But that's my opinion.
- About the `Sort` option, to me, all sorting, filtering and paging options should be tied to a dataset for server-side operations. And for client side sorting, it's not enough to have only one method because each element in treeView could have its own type (string, number, date...). So i think, each data element could implement their own sorting function depending on their types or have a custom sorting function that the dev can set themselves.
- About the `Lazyloading` attribute, i think it is not necessary. For server side, if you implement a dataset interface you can specify this kind of option by setting the offset/limit attribute for instance, and by default, the server shouldn't send you back all children, but only if you ask for one specific leaf/branch. All of that, should be tied into the dataset interface. And you could use viewport virtualisation in order to optimise the component memory consumption. I've made the virtual scroller already (i've made a fork of https://rintoj.github.io/ngx-virtual-scroller/demo and optimise it a bit more). I'm using it for my datagrid already.
- All parameters would need to be camel-cased in their implementation.

Let me know what you think guys :-D

**Comment by JoMas971** [comment]
This is my first thought concerning the directory interface : 

```typescript
// A root can have multiple branches and/or leafs
// A branch can have multiple leafs
// A leaf... is a leaf, what did you expect ?
enum TreeViewDirectoryTypeEnum {
  ROOT = 'root',
  BRANCH = 'branch',
  LEAF = 'leaf'
}

// This is what i expect to see for the directories.
interface TreeViewDirectoryInterface {
  data: any; // The raw data object.
  field: any; // The identifier to retrieve the value. It should be something like value = data[field] behind the scene.
  value: any; // The raw value set behind the scene.
  getValue: () => any; // This will return the formatted value and used in template.
  valueFormatter: () => any; // This function will be used to format the value.
  type: TreeViewDirectoryTypeEnum;
  selected: boolean; // Whether or not it's selected
  
  onFocus?: () => void; // Bind the click event to this in case the user has some extra stuffs to do on his side
  onBlur?: () => void;
  expanded?: boolean; // Whether or not it's expended or collapsed.
  sortType?: string | number | 'date' | 'custom'; // What kind of data we have for sorting operations.
  loadChildren?: (value) => void; // Only for server-side operations. It will be called when the user expand a branch.
                                  // It will use the defined ServerSideInterface behind the scene to do the call. 
  parent?: TreeViewDirectoryInterface;
  children?: TreeViewDirectoryInterface[];
}
```

**Comment by cfsnyder** [comment]
**Tree View API**
```typescript
class TreeViewComponent<T> {

  constructor(private rootNode: TreeNode<T>, private config: TreeViewConfig) {}

  public getEventObservable(): Observable<TreeViewEvent<T>> {
    ...
  }

  public selectNode(node: TreeNode<T>) {
    ...
  }

  public expandNode(node: TreeNode<T>) {
    ...
  }

  public collapseNode(node: TreeNode<T>) {
    ...
  }

  public toggleNode(node: TreeNode<T>) {
    ...
  }

}

interface TreeViewEvent<T> {
  getNode(): TreeNode<T>;
  getType(): TreeViewEventType;
}

enum TreeViewEventType {
  NODE_CLICKED,
  NODE_EXPANDED,
  NODE_COLLAPSED,
  ...
}

interface TreeNode<T> {
   getLabel(): string;
   getData(): T;
   hasParent(): boolean;
   hasChildren(): Promise<boolean>;
   getParentNode(): TreeNode<T> | null;
   getChildNodes(): Promise<Array<TreeNode<T>>>;
}

interface TreeNodeData<T> {
  data: T;
  label: string;
}

interface TreeViewDataRetriever<T> {
  
  getChildNodeData(parent: TreeNode<T>): Promise<Array<TreeNodeData<T>>;

  hasChildNodes(parent: TreeNode<T>): Promise<boolean>;

}

class TreeNodeImpl<T> implements TreeNode<T> {
  
  constructor(private data: TreeNodeData<T>,  private dataRetriever: TreeViewDataRetriever) {}

  ...  // omitting other interface methods

  getChildNodes(): Promise<Array<TreeNode<T>> {
   return this.dataRetriever.getChildNodeData(this).then(childData => {
    return childData.map(it => new TreeNodeImpl(it, this.dataRetriever);
  });

  ...

 }
```

Someone integrating this just needs to provide the root node `TreeNodeData` and the `TreeViewDataRetriever`. You can imagine that the synchronous implementation of the `TreeViewDataRetriever` would just always return immediately resolved promises with the requested data.

Obviously I still left a few details out that will need to be defined prior to implementation.

**Comment by cfsnyder** [comment]
One thing that occurred to me that we'll need to handle in the visual design with @mathieupkf  is the presentation of errors when when of the promises returned from the `TreeViewDataRetriever` is rejected (due to a network/server error, or otherwise).

**Comment by JoMas971** [comment]
Thanks @cfsnyder for your input. That look pretty good ! 
I just have a question about how many nodes we should get in a call. 
For instance, if i want to expand a node that have 1000 children Server Side. How should i "lazy-load" them with your solution If for instance i want to get chunks of 10 items before loading more ?

**Comment by cfsnyder** [comment]
@JoMas971 you just extend the models to include new implementations that support paging features. I've updated my above comment with those extensions included.

**Comment by mathieupkf** [comment]
@JoMas971 @cfsnyder @ariquinones 
I've created a new screen for our Treeview Component. I've designed several cases. Please let me know if you have improvements in mind or other variations needed. 
https://projects.invisionapp.com/d/main#/console/12026582/383168113/preview

**Comment by JoMas971** [comment]
thanks @mathieupkf , that looks good ! So if i follow you, it would be possible to use a custom template to handle the "arrows" or "+/-" buttons right ? And the template for each line should be customisable to please whatever the dev want to do right ? That should be pretty easy to do using Angular `TemplateRef<any>` and `<ng-container>`...
Example of those can be found within the datagrid component.
Also, concerning the design, it should be easy to override the colors in order to quickly make a new theme. That would be easy using SCSS we just need to know from the beginning which parts could be coloured etc... Which is the case with your design.
But for the first push, we would have only one colour theme (the one we will use along with the datagrid). We would add more later on when we need them.

**Comment by ariquinones** [comment]
@cfsnyder and @JoMas971 I’ve made some progress on the Tree View component and added some more stuff to the API. Everything from Cory’s above comment has stayed the same while adding the following to some of the interfaces:
You can see the full code in my fork branch:
https://github.com/ariquinones/ferui/tree/topic/tree/src/ferui-components/tree-view

And how to use it by a developer would be in the demo part:
https://github.com/ariquinones/ferui/tree/topic/tree/src/dev/src/app/components/tree-view
      
        // Added the getLabel() and getIcon() methods
        interface TreeNode<T> {
          getLabel(): string; // to get the label/field we need to get and display from the data object
          getIcon(nodeExpanded: boolean, nodeSelected: boolean): TemplateRef<any> | null // this is to allow the developer to chose what icon they would like to be present next to the node. Default is no icon.
        }

        interface TreeNodeDataRetriever<T> {
          getIconTemplate(node: TreeNode<T>, expanded: boolean, selected: boolean): TemplateRef<any> | null // get the Icon for each node according to its current state
        }

        interface TreeNodeData<T> {
          limit?: number | ‘*’; // used on Server Side rendering to know the params limit, can take in * for unlimited
        }

        // Config object for a developer to explicitly choose the Tree View width and height, if nodes should have borders like in Mathieu’s design
        // And choose a color variation theme from what Mathieu provided or choose their own color scheme
        // All accepts standard CSS color values such as hex values, color names, and rgb. If nothing is given we’ll use the WHITE default color
        interface TreeViewConfiguration {
          width?: string;
          height?: string;
          hasBorder?: boolean;
          colorConfiguration?: TreeViewColorConfiguration;
          colorVariation?: TreeViewColorVariation;
        }
        
        type TreeViewColorVariation = 'DARK_BLUE' | 'LIGHT_BLUE' | 'WHITE';
	
        interface TreeViewColorConfiguration {
          backgroundColor: string;
          textColor: string;
          nodeEventColor: NodeEventColor;
        }

        interface NodeEventColor {
          backgroundHoverColor: string;
          backgroundSelectColor: string;
          textSelectColor: string;
        }

@mathieupkf One thing we'll need to add to the design would be the case where we only load a limited amount of children for a parent node, we'll need to add some sort of "view more children" button in order to load the next batch under that parent node. This action button will need to be the same overall height as the rest of the nodes please 

Still need to work out the details for implementing this api with a virtual scroller

**Comment by mathieupkf** [comment]
@ariquinones - New design version for the View More case. You'll tell me when (and how) I'll be able to test the component! Thanks. 
https://projects.invisionapp.com/d/main#/console/12026582/383168113/preview

**Comment by JoMas971** [comment]
@ariquinones thanks for the update ! 
I just have some opinions about some of the stuff you gave us : 
- First, i don't see the point of adding the `getIcon(...)` method on `TreeNode<T>` interface. This sort of things should be handled by the template itself. If you want to provide the option to choose from an existing kind of icon, i suggest to only add one option to the configuration object to choose the kind of icon from a list that you'll build. 
Something like : 
```
{
  ...,
  iconType: TreeNodeIcons.ARROW,
  ....
}
```
Also, the `getIconTemplate(...)` method shouldn't be only for changing the icon. It should be `getTemplate(...)` instead and could be overridden by the developper if he want to do special rendering stuffs (changing the icon and use one of his own, or change the order of icon/text, or if he want to add extra stuffs... etc), but there should be a default template that render the entire node (icon + text) if the dev doesn't want to change the rendering of each nodes.
- Secondly, i don't think we should handle a color theme within the component itself. All design should be handle in `ferui-design` project, same apply to theming. And if you want to apply a specific theme to your component, it would be through CSS classes. If you want to let the user choose a specific theme, you can indeed list all available color-themes (create a string enum and map the corresponding classes for each theme) and add a configuration option to let the user choose his theme from the existing color-themes OR add his own specific CSS class (if he want to override it and create his own theme).
But since theming is not the most important thing to do at first, i suggest to delay this a bit and have only one theme at first (the one @mathieupkf want in his treeview + datagrid component). 
We want a working, fully tested, treeview component ASAP so don't bother with adding multiple theme now...

@mathieupkf thank you for updating the design. I just have one question : Why did you added the "children" or "parents" words after the 'view more' text button ? 
A node can be a parent of another node but still being a child of another parent... So it doesn't make sense to add those extra words... But maybe you want to use the name of the node instead ? In that case, we will have some issues with long names...
Let say i'm in a folder/files treeview configuration, and my folder is named 'All stuffs that i don't want for my daughter', if we do this solution we would have something like : 

- All stuffs that i don't want for my daughter
     - Boys
     - Sickness
     - (...)
     - **view more of All stuffs that i don't want for my daughter**

I think we can just use indentation to separate one 'view more' button from another one and only display the 'view more' text :-).

What do you think @cfsnyder @ariquinones @mathieupkf ?

**Comment by cfsnyder** [comment]
@ariquinones I'm not understanding why we need a "view more" button. Shouldn't the virtual scroll be responsible for loading more lazily, in the background.

**Comment by cfsnyder** [comment]
@ariquinones the `limit` parameter on the TreeNodeData interface doesn't make sense to me. Can you explain?

**Comment by cfsnyder** [comment]
I agree with the first two points made by @JoMas971. Don't worry about providing specific customization options for styles, instead just let the developer completely override the HTML template that is used for an individual node and let them style it however they like.

**Comment by cfsnyder** [comment]
@ariquinones does `TreeNode.getLabel` just proxy `TreeNode.getData().label`?

**Comment by JoMas971** [comment]
@cfsnyder About the 'view more' button, i told him to do so because it is not a good idea to do infinite scrolling for this kind of component, because it's a hierarchical construction.
For instance, consider this use case (i'll be in a folder/file configuration) : 

- All stuffs that i don't want for my daughter
    - Boys
    - Sickness
    - (1000 more items are listed here...)
- All stuffs that my boy did with me
    - Biking
    - Swiming
    - (100 more items loaded there)
- (and so on...)
    - (you've got the point...)

If i open the first folder then scroll down to the bottom, i'll wait ALL 1002 items to be load (through infinite scrolling) before being able to see my second folder... Also, that will be a lots of un-needed api call to fetch the data...
At a UX point of view, it can't work with treeview. That's why i've told him to use this 'view more' option. But we will have a large limit by default (like 50 i think) so it wouldn't see that button too much.

BTW, the virtual scroller will only virtualise an array of objects in order to have fewer elements loaded in the DOM. It doesn't handle infinite scrolling.

**Comment by ariquinones** [comment]
@JoMas971 thanks for the feedback! Yup you're right about the getIcon() method, I'll need to switch that to a getTemplate() method to let the developer give us a template for the entire node (text + icon). If nothing is provided I'll use the default non-icon template. 
Ah okay, yes I'll update the color theme config to use classes like you suggest and we'll let the developer overwrite them how Cory suggests as well. 
@cfsnyder just how Jon mentioned the view more button will be needed because from a hierarchical stand point it would be impossible to detect what node the user is truly scrolling and wanting to load more children off. 
The limit parameter is on the TreeNodeData since that'll pertain to each individual node when using server side, where else would you recommend putting that in?
Yes the getLabel is just a proxy to get the label field on the data for an easier get method on the template. 
And yes, I'll let the developer overwrite the css like you mentioned!

**Comment by cfsnyder** [comment]
@ariquinones - I may be missing something but the assertion that "it would be impossible to detect what node the user is truly scrolling and wanting to load more children off" is hard for me to understand. From my perspective, the user is never scrolling a particular node.. they're always scrolling the entire tree view. If you know the height of individual nodes in the display and you know the number of children for a particular expanded node (which is part of the paged API), you should be able to know which ones need to be in in view at any particular time. Am I wrong?

**Comment by cfsnyder** [comment]
@ariquinones - I don't see how the `limit` parameter is at all relevant there, you'll have to explain how exactly you're intending to use it?

**Comment by JoMas971** [comment]
@ariquinones i agree with @cfsnyder about the limit param. It shouldn't be there. It's already part of `PagingParams` interface. You can assume that if the user set a limit to -1, that means 'no-limit'. So maybe the '*' notation wasn't a good idea after all... Sorry about that.

@cfsnyder we wouldn't have multiple scrolls by nodes indeed, but what Ari has trying to say is that, when you open a node, then scroll down, if you set an infinite scroll there, and if your node contains a lot of other nodes (more than 100 nodes), it will be bad for the user if, for instance, he's trying to scroll to the next node instead.
He'll need to load ALL previous nodes (knowing the amount or not) before going to the node that he actually want to go to. Does that make sense ?

The only solution for the user if we implement an infinite scrolling will be to collapse the node he doesn't want to load children from, then go to the node he want. But at a UX point of view, i'm not sure it's the best thing to do...

@mathieupkf have you ever seen treeviews systems that implement infinite scrolling ? Maybe you can give us your thoughts about this :-).

**Comment by cfsnyder** [comment]
@JoMas971  - what you're saying does make sense to me but I'm not convinced that we need a "load more" vs. just letting them collapse the node with many children that is irrelevant to what they're looking for. I don't recall having seen a "load more" option in a tree view before, so that makes me think it probably isn't necessary.

Not sure that is what @ariquinones  was trying to convey, though? Seems that he was suggesting it was impossible to implement vs. a matter of improving UX?

**Comment by JoMas971** [comment]
I'll let @ariquinones defend himself about what he was trying to say then lol.

On my side, i've never saw a treeview system that implement a true infinite scrolling (i mean by that, loading nodes in chunks with some limits depending on where the scrollbar is) :-/ 
So i was thinking that it wasn't out there for a reason... :-)

The only implementation i've found of an "infinite scroll treeview" was this : https://infinite-tree.js.org/examples.html#/classic
But you can't set any limit there... Also, it will load ALL children when you expand a node. So at a UX point of view it's a bit better even though it won't fix the issue i was talking about which is : Open a node that has a huge amount of nodes, but trying to go to the next direct node instead... You'll need to scroll for a while (in their examples, they load 10k nodes...). 

But, if it make sense to you to have that (instead of our 'view more' solution which is more optimised) we can definitely implement something similar. 
But, to be clear, it has nothing to do with the virtual scroller. We'll need to implement a custom infinite scrolling solution for treeview, like i did with our current infinite scrolling row model in datagrid ;-). But that will be a bit more complicated...

**Comment by JoMas971** [comment]
this issue has been fixed by #54 .
I'll close the issue.
